### PR TITLE
Cleanup character io interfaces

### PR DESCRIPTION
**Pull request by rmkaplan**
I think this is ready for more general testing and eventual merging.  It isolates the XCCS/NS external format dependencies into one new file XCCS, and all character access then goes through a few abstract functional interfaces (\INCCODE, \PEEKCCODE...), with the XCCS implementations still set up as the default.

This required touching and recompiling lots of files, but it all seems to hang together in this full sysout and in the LFG system built on top.

Other improvements/fixes to MODERNIZE, WHEELSCROLL, CLIPBOARD, and others.

### CODE REVIEW

**Comment by rmkaplan** [comment]
I have updated to fix some glitches that I observed in further testing.  In particular, the interface to MAKEFILE now lets you specify a new format (like :UTF-8), and that propagates through remaking and also through compiling and recompiling.

Those are new capabilities.  The main strategy for reviewing and testing this branch would be just to run in it.  All old functionality is supposed to work as it used to, despite the changes to low-down interfaces. That's the first confirmation.

It may be appropriate to merge when all old behavior appears to be stable and correct.  Then we can inch up on the new encoding capabilities.

**Review by masinter — COMMENTED** [review]
why is sources/LLREAD.LCOM.~1~ on .gitignore? it already ignores *~

**Review by masinter — COMMENTED** [review]
there were changes (introductions of functions) to BIGBITMAPS in the library -- not loaded in FULL. Were these not part of the IO functions?

**Review by masinter — COMMENTED** [review]
CHAT was recompiled with no changes. CHAT(SHELL) is the only useful CHAT target.
But shouldn't it assume :UTF-8 ?   I think CHAT(host) could turn into ssh but for now it is better to leave out?
CHAT(SHELL) currently always opens /bin/csh (C-shell) which isn't in most Linux distros. See #121

**Comment by rmkaplan** [comment]
I think this was something else that got included to make the masterscope analysis more accurate, making sure that all ambiguous field names were qualified.

**Comment by rmkaplan** [comment]
Could be that macros or records changed underneath.

**Comment by rmkaplan** [comment]
Dunno

**Review by masinter — COMMENTED** [review]
there was a change to SYSEDIT that removed resetting MSMACROPROPS to leave MACRO-FN on the list, leaving the value on MSANALYZE. Macros for all kinds are analyzed by analyzing their expansion. There was also a patch to remove / no longer generate spurious incorrect masterscope "templates" (call to SETTEMPLATE in GETTEMPLATE).
Otherwise,

File: .gitignore [hunk 1]
```diff
@@ -27,3 +27,4 @@ core
 
 # set up by install-diff-filter.sh script
 .gitattributes
+sources/LLREAD.LCOM.~1~
```
Comment (masinter) [inline .gitignore:30]: *~ is already gitignored

File: library/BIGBITMAPS [hunk 2]
```diff
@@ -1,34 +1,38 @@
...
+      |changes| |to:|  (FNS \\DRAWLINE.BIGBM.DASH \\DRAWLINE.BIGBM.NODASH BIGBITMAPP)
+                       (VARS BIGBITMAPSCOMS)
 
-      |previous| |date:| "20-Jan-93 13:40:44" |{PELE:MV:ENVOS}<LISPCORE>LIBRARY>BIGBITMAPS.;10|)
+      |previous| |date:| "10-May-2021 15:37:51" 
+|{DSK}<Users>kaplan>Local>medley3.5>git-medley>library>BIGBITMAPS.;1|)
 
 
-; Copyright (c) 1991, 1993, 1994 by Venue.  All rights reserved.
+; Copyright (c) 1991, 1993-1994, 2021 by Venue.
```
Comment (masinter) [inline library/BIGBITMAPS:12]: new copyright could say Interlisp.org

File: library/BIGBITMAPS [hunk 3]
```diff
@@ -1,34 +1,38 @@
...
-       ((DECLARE\: DONTEVAL@LOAD EVAL@COMPILE DONTCOPY (RECORDS BIGBM)
+       ((DECLARE\: EVAL@COMPILE DONTCOPY (RECORDS BIGBM)
                (CONSTANTS (|\\MaxBitMapHeight| 65535)
                       (|\\MaxBitMapWidth| 65535)
                       (|\\MaxBitMapWords| 131066))
                (MACROS |GetNewFragment|)
                (MACROS |\\SFInvert|))
         (INITRECORDS BIGBM)
-        (FNS BITBLT.BIGBM BITMAPCREATE.BIGBM BITMAPCREATE BITMAPCOPY BLTSHADE.BIGBM BITBLT 
+        (FNS BIGBITMAPP BITBLT.BIGBM BITMAPCREATE.BIGBM BITMAPCREATE BITMAPCOPY BLTSHADE.BIGBM BITBLT
```
Comment (masinter) [inline library/BIGBITMAPS:24]: is there any documentation for these changes or BIGBITMAP itself?

File: library/BIGBITMAPS [hunk 4]
```diff
@@ -65,11 +69,11 @@
 
 (PUTPROPS |\\SFInvert| MACRO ((|BitMap| \y)
 
-         (* |corrects| |for| |the| |fact| |that| |alto| |bitmaps| |are| |stored| |with|
```
Comment (masinter) [inline library/BIGBITMAPS:68]: can we convert these comments to strings? Or use a different readtable than XCL?

File: library/BIGBITMAPS [hunk 5]
```diff
@@ -84,6 +88,10 @@
        '6)
 (DEFINEQ
 
+(BIGBITMAPP
```
Comment (masinter) [inline library/BIGBITMAPS:91]: I suppose these changes are to avoid the cross-file record access. Sorry that just became clear.

**Comment by rmkaplan** [comment]
Is this a change that didn’t get captured in this branch?  I don’t remember doing anything about macroprops or templates.

I did fix masterscope to go to the whereis database as a last resort for CONTAINS queries, so I could find the things I needed to work on.

(Also, I extended DUMPDATABASE to take a FILE argument so that it could dump the database with a define-file-info reader environment that would load in all contexts.)

**Comment by rmkaplan** [comment]
Yes, to export a test that didn’t require the declaration to be present.  (The logic in that code is still a little confusing, I think the bigbitmap extension is wired in at a slightly wrong level.  But I left it.)

File: lispusers/MODERNIZE [hunk 6]
```diff
@@ -1,11 +1,11 @@
-(DEFINE-FILE-INFO READTABLE "INTERLISP" PACKAGE "INTERLISP")
-(FILECREATED "14-Mar-2021 20:33:34" 
-{DSK}<Users>kaplan>Local>medley3.5>git-medley>lispusers>MODERNIZE.;14 20950
```
Comment (masinter) [inline lispusers/MODERNIZE:3]: shouldn't this move to Library? Basically anything that is part of FULL.

**Review by masinter — CHANGES_REQUESTED** [review]
The problem with MSMACROPROPS is just an indication that you made your branch and didn't update it from master for months; I wasn't doing a _lot_ of editing but we need to resolve the conflicts (not many).

I've gone through about 50 / 150 of the changed files in about 5 hours. I'll try to go through the rest.

**Comment by rmkaplan** [comment]
Thanks for slogging through this, I know it’s a lot.

In terms of legacy dates, I thought it was set up to do that.  It certainly does that in my LFG environment.  Maybe there is a flag some where that should be set.

In fact, it seems that in LFG I have a file EDITDATEPATCH that has a variable SINGLEEDITDATE that controls that.  Maybe the system edit date isn’t set up that way.

File: sources/CMLREAD [hunk 7]
```diff
@@ -237,38 +240,41 @@ Copyright (c) 1985, 1986, 1987, 1988, 1990, 1993 by Venue & Xerox Corporation.
...
+       '10)
 
 (DEFMACRO WITH-READER-ENVIRONMENT (ENV . BODY)
    `((CL:LAMBDA (E)
             (LET ((*PACKAGE* (ffetch (READER-ENVIRONMENT REPACKAGE) of E))
                   (*READTABLE* (ffetch (READER-ENVIRONMENT REREADTABLE) of E))
                   (*READ-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E))
-                  (*PRINT-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E)))
+                  (*PRINT-BASE* (ffetch (READER-ENVIRONMENT REBASE) of E))
+                  (*EXTERNALFORMAT* (ffetch (READER-ENVIRONMENT REFORMAT) of E)))
```
Comment (masinter) [inline sources/CMLREAD:260]: where are the other variables declared SPECIAL? Why just the external format? Is this really at the right level? The other elements don't depend / vary when you change XCCS and UTF-8... they're EXTERNAL-FORMAT independent

File: sources/CMLREAD [hunk 8]
```diff
@@ -237,38 +240,41 @@ Copyright (c) 1985, 1986, 1987, 1988, 1990, 1993 by Venue & Xerox Corporation.
...
-(PUTPROPS WITH-READER-ENVIRONMENT INFO EVAL)
+(PUTPROPS WITH-READER-ENVIRONMENT INFO EVAL)
 (DECLARE%: DOEVAL@COMPILE DONTCOPY
 
 (GLOBALVARS *COMMON-LISP-READ-ENVIRONMENT*)
 )
 
 (RPAQ? *COMMON-LISP-READ-ENVIRONMENT* (create READER-ENVIRONMENT REPACKAGE _ (CL:FIND-PACKAGE
                                                                                   "USER")
-                                                 REREADTABLE _ CMLRDTBL REBASE _ 10))
```
Comment (masinter) [inline sources/CMLREAD:269]: same ... common lisp vs. interlisp readtable shouldn't change the "stream" properties.

File: sources/EXTERNALFORMAT.TXT [hunk 9]
```diff
@@ -0,0 +1,70 @@
+New architecture for character input-output and alternative external formats
```
Comment (masinter) [inline sources/EXTERNALFORMAT.TXT:1]: we really need to put together a Medley 4 release manual, formatted and fixed up

File: sources/FASLOAD [hunk 10]
```diff
@@ -1,14 +1,14 @@
 (DEFINE-FILE-INFO READTABLE "XCL" PACKAGE "FASL")
-(IL:FILECREATED "17-Apr-2018 07:55:20" 
-IL:|{DSK}<Users>kaplan>Local>medley3.5>lispcore>sources>FASLOAD.;2| 35249  
+(IL:FILECREATED "10-Jun-2021 18:26:43" 
+IL:|{DSK}<Users>kaplan>Local>medley3.5>git-medley>sources>FASLOAD.;8| 35584  
 
-      IL:|changes| IL:|to:|  (IL:FNS CONVERT-FASL-DATE)
```
Comment (masinter) [inline sources/FASLOAD:5]: This is really illegible IL:|changes| IL:|to:|

**Review by masinter — CHANGES_REQUESTED** [review]
while there are a lot of "cosmetic" issues (bad pretty printing, edit comments lost, etc) there are a few more serious issues.  I think it's necessary to sort out the "Masterscope gives error on ambiguous record field" changes from the "support external character encoding". The former could be fixed by adding a per-file record priority list, removing the necessity of any source edits. And the READTABLE should be independent of the file encoding. READ works on a stream of characters, which remain the same character independent of whether they are encoded in XCCS, UTF-8, ISO8059-10 or what have you.
I propose using this branch as a model, and bringing forward changes one at a time starting with current master.

File: sources/FASLOAD (hunk 10 above)
Comment (nbriggs) [reply sources/FASLOAD:5]: Happens any time the FILECREATED expressions are written/read with a readtable other than an IL based one.  IL:|previous| IL:|date|, etc.

**Comment by rmkaplan** [comment]
With respect to CMLREAD, the other variables are the variables that define the current default readtabe, package, and base, when you call IO functions with those not locally specified.  I presume that there are global declarations somewhere that they are specvars.

I added the external format as another variable, when I thought that it was of the same sort and that was the right thing to do.  But it isn’t, the external format is a property of the stream itself, not of the printing environment.  So that variable is really not interpreted—the external format is installed specifically when the stream is opened or the user changes it by STREAMPROP or SETFILEINFO.

To make it operate implicitly as part of the context set by this macro, a further change would have been to have this reset the format of the stream under a resetsave and restore the previous external format on exit.  But the problem is that this macro doesn’t actually take the stream as an argument, it just sets these variables for any and all printing/reading to any and all streams underneath.

(Note also that setting the format changes all 5 of the primitive character reading/writing functions in the stream. It’s a level below whether or not a given internal character is a sepr in this context, or a package identifier should be printed.)

(Note also:  I have no idea why this was in CMLREAD)

**Comment by rmkaplan** [comment]
Also, the externalformat.txt document reflects an earlier stage of the migration from what I inherited and where I ended up.  I’ll do an update when this settles down.  (After this merge, there are a few minor cleanups (rationalizing a few of the naming conventions, maybe even taking the format variable out of the WITH-READER-ENVIRONMENT) still to do.)

**Comment by masinter** [comment]
Can we at least (a) implement some "unit tests" that makes sure the UTF-8 code is working? Preferably something that can be automated? 
(b) agree to make up "issues" that cover the problems I identified, as a way of getting agreement on what problems you agree need fixing and why? 
I didn't finish reviewing all of the files, and those that diff well (because of pretty printing) are an issue

File: sources/FASLOAD (hunk 10 above)
Comment (masinter) [reply sources/FASLOAD:5]: a separate issue, but I think we should change the FILECREATED format to use strings or keywords

   "changes to" IL:READ "previous" "12-DEC-2021"

or

   :changes "IL:READ" :previous-date "12-Dec-2021"

so it isn't so ugly. This is in support of "Modernization" and "better support of Common Lisp".

File: .gitignore (hunk 1 above)
Comment (masinter) [reply .gitignore:30]: now issue #357

File: library/BIGBITMAPS (hunk 2 above)
Comment (masinter) [reply library/BIGBITMAPS:12]: Issues #207 (copyright dates) and (possibly?) #39 (larger screen bitmap)

File: library/CLIPBOARD [hunk 11]
```diff
@@ -203,8 +203,8 @@
 )
 (PUTPROPS CLIPBOARD COPYRIGHT (NONE))
```
[the remark this answers is not in the document]
Comment (masinter) [reply library/CLIPBOARD:204]: See notes in #211

File: library/POSTSCRIPTSTREAM [hunk 12]
```diff
@@ -1113,59 +1111,63 @@ Copyright (c) 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1997, 1998, 2018 by Venu
...
+                                                                   \PS.SCALE0)
+                                                            T))
                 "%%%%Title: "
                 (MKSTRING (OR (LISTGET OPTIONS 'DOCUMENT.NAME)
                               FILE))
                 T "%%%%Creator: PostScript Driver Copyright (C) 1988-1992 Venue and others" T 
                 "%%%%CreationDate: " (DATE)
                 T %# (COND
-                        ((EQ 'LPT (FILENAMEFIELD FP 'HOST))
+                        ((EQ 'LPT (FILENAMEFIELD STREAM 'HOST))
```
[the remark this answers is not in the document]
Comment (masinter) [reply library/POSTSCRIPTSTREAM:1162]: #363 moderrnize printing

File: lispusers/ISO8859IO [hunk 13]
```diff
@@ -35,23 +35,10 @@ Copyright (c) 1995-1997, 1999, 2021 by Xerox Corporation.
...
-               (CONSTANTS (NSCHARSETSHIFT 255))
-               
-               (* ;; "From LLREAD")
-
-               (MACROS \XCCSIN \XCCSPEEK \BACKXCCSCHAR)
-               
-               (* ;; "From MODARITH")
-
-               (MACROS UNFOLD))))
+        (DECLARE%: EVAL@COMPILE DONTCOPY (P (OR (GETP 'EXPORTS.ALL 'FILE)
```
[the remark this answers is not in the document]
Comment (masinter) [reply lispusers/ISO8859IO:38]: #365

File: lispusers/MODERNIZE [hunk 14]
```diff
@@ -92,19 +101,28 @@
...
-        (WINDOWPROP WINDOW 'BUTTONEVENTFN (IF ANYWHERE
-                                              THEN (FUNCTION MODERNWINDOW.BUTTONEVENTFN.ANYWHERE)
-                                            ELSE (FUNCTION MODERNWINDOW.BUTTONEVENTFN))))
+    (CL:WHEN (AND TITLEPROPORTION (GREATERP TITLEPROPORTION 0.5))
+           (ERROR "TITLEPROPORTION cannot be greater than .5"))
+    (CL:WHEN (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN)
+        (WINDOWPROP WINDOW 'BUTTONEVENTFN (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN))
+        (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN NIL))
+    (WINDOWPROP WINDOW 'PREMODERN-BUTTONEVENTFN (WINDOWPROP WINDOW 'BUTTONEVENTFN))
+    (WINDOWPROP WINDOW 'BUTTONEVENTFN (IF (OR ANYWHERE TITLEPROPORTION)
```
[the remark this answers is not in the document]
Comment (masinter) [reply lispusers/MODERNIZE:116]: issue #367

File: sources/CMLREAD (hunk 7 above)
Comment (masinter) [reply sources/CMLREAD:260]: Issue #371

**Comment by masinter** [comment]
no longer needed; was turned to draft after merging to make sure issues raised / tested

**Pull request closed without merging**